File: app/src/main/java/com/example/merchio/adapters/CategoryAdapter.java
package com.example.merchio.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.example.merchio.R;
import com.example.merchio.models.Category;

import java.util.List;

public class CategoryAdapter
        extends RecyclerView.Adapter<CategoryAdapter.ViewHolder> {

    Context context;
    List<Category> categoryList;

    public CategoryAdapter(Context context,
                           List<Category> categoryList) {

        this.context = context;
        this.categoryList = categoryList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(
            @NonNull ViewGroup parent,
            int viewType) {

        View view = LayoutInflater.from(context)
                .inflate(
                        R.layout.item_category,
                        parent,
                        false
                );

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(
            @NonNull ViewHolder holder,
            int position) {

        Category category =
                categoryList.get(position);

        holder.txtCategory.setText(
                category.getName()
        );

        Glide.with(context)
                .load(category.getIconUrl())
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .skipMemoryCache(true)
                .into(holder.imgCategory);
    }

    @Override
    public int getItemCount() {
        return categoryList.size();
    }

    public static class ViewHolder
            extends RecyclerView.ViewHolder {

        ImageView imgCategory;
        TextView txtCategory;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            imgCategory =
                    itemView.findViewById(R.id.imgCategory);

            txtCategory =
                    itemView.findViewById(R.id.txtCategory);
        }
    }
}

File: app/src/main/java/com/example/merchio/adapters/CategoryAdapter.kt
package com.example.merchio.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.merchio.R
import com.example.merchio.data.model.CategoryDto

class CategoryAdapter(private val onClick: (CategoryDto?) -> Unit) : RecyclerView.Adapter<CategoryAdapter.CategoryViewHolder>() {
    private val items = mutableListOf<CategoryDto?>()
    private var selectedIndex = 0

    fun submitList(data: List<CategoryDto>) {
        items.clear()
        items.add(null)
        items.addAll(data)
        selectedIndex = 0
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        return CategoryViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_category, parent, false))
    }

    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) = holder.bind(items[position], position)
    override fun getItemCount(): Int = items.size

    inner class CategoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val name: TextView = itemView.findViewById(R.id.txt_category_name)
        fun bind(item: CategoryDto?, position: Int) {
            name.text = item?.name ?: "All"
            itemView.isSelected = position == selectedIndex
            itemView.setOnClickListener {
                selectedIndex = bindingAdapterPosition
                notifyDataSetChanged()
                onClick(item)
            }
        }
    }
}
